### PR TITLE
[CARBONDATA-3297] Fix that the IndexoutOfBoundsException when creating table and dropping table are at the same time

### PR DESCRIPTION
**Pull request by qiuchenjian**
### [Problem]
Throw  the IndexoutOfBoundsException when creating table and dropping table are at the same time
`java.lang.IndexOutOfBoundsException: 179
        at scala.collection.mutable.ResizableArray$class.apply(ResizableArray.scala:43)
        at scala.collection.mutable.ArrayBuffer.apply(ArrayBuffer.scala:48)
        at scala.collection.IndexedSeqOptimized$class.segmentLength(IndexedSeqOptimized.scala:195)
        at scala.collection.mutable.ArrayBuffer.segmentLength(ArrayBuffer.scala:48)
        at scala.collection.GenSeqLike$class.prefixLength(GenSeqLike.scala:93)
        at scala.collection.AbstractSeq.prefixLength(Seq.scala:41)
        at scala.collection.IndexedSeqOptimized$class.find(IndexedSeqOptimized.scala:50)
        at scala.collection.mutable.ArrayBuffer.find(ArrayBuffer.scala:48)
        at org.apache.spark.sql.hive.CarbonFileMetastore.getTableFromMetadataCache(CarbonFileMetastore.scala:203)
        at org.apache.spark.sql.CarbonEnv$.getCarbonTable(CarbonEnv.scala:203)
        at org.apache.spark.sql.CarbonEnv$.getTablePath(CarbonEnv.scala:288)
        at org.apache.spark.sql.execution.command.table.CarbonCreateTableCommand$$anonfun$1.apply(CarbonCreateTableCommand.scala:74)
        at org.apache.spark.sql.execution.command.table.CarbonCreateTableCommand$$anonfun$1.apply(CarbonCreateTableCommand.scala:74)`

### [Solution]

1. The type of carbonTables in MetaData.class is ArrayBuffer, and the ArrayBuffer is not thread-safe, so it  throw this exception when creating table and dropping table are at the same time

2. Use read write lock to guarantee the thread-safe

 

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2585/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10845/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2816/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2586/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10846/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2817/

**Comment by xuchuanyin** [comment]
@ravipesala @brijoobopanna 
I think concurrently manipulating the metastore is quite a fundermental usecase and **should not have** problems...

**Comment by xuchuanyin** [comment]
@ravipesala @brijoobopanna 
I think concurrently manipulating the metastore is quite a fundermental usecase and **should not have** problems...

**Comment by qiuchenjian** [comment]
@ravipesala  @brijoobopanna @xuchuanyin   
This exception is throwed by create table command by chance， not every time.
I think the reason is  type of carbonTables in MetaData.class is ArrayBuffer, and the ArrayBuffer is not thread-safe and we drop tables and create tables at the same time in a carbon session 
Do you have other ideas?

**Comment by akashrn5** [comment]
@qiuchenjian i think red lock should be added for `listAllTables` APi also, please check it once

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala [hunk 1]
```diff
@@ -19,6 +19,7 @@ package org.apache.spark.sql.hive
 
 import java.io.IOException
 import java.net.URI
+import java.util.concurrent.locks.{Lock, ReadWriteLock, ReentrantReadWriteLock}
```
Comment (akashrn5) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala]: remove unnecessary imports

Comment (qiuchenjian) [reply integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala]: done

**Comment by qiuchenjian** [comment]
> @qiuchenjian i think red lock should be added for `listAllTables` APi also, please check it once

done!
I also clone a new seq, besause it's useless that return  the same ArrayBuffer

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2671/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10930/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2899/

**Comment by akashrn5** [comment]
> > @qiuchenjian i think red lock should be added for `listAllTables` APi also, please check it once
> 
> @akashrn5
> done!
> I also clone a new seq, besause it's useless that return the same ArrayBuffer
> Please review and merge, thank you

@qiuchenjian sorry i did not get this `besause it's useless that return the same ArrayBuffer`

**Comment by qiuchenjian** [comment]
> > > @qiuchenjian i think red lock should be added for `listAllTables` APi also, please check it once
> > 
> > 
> > @akashrn5
> > done!
> > I also clone a new seq, besause it's useless that return the same ArrayBuffer
> > Please review and merge, thank you
> 
> @qiuchenjian sorry i did not get this `besause it's useless that return the same ArrayBuffer`

@akashrn5 
I mean that if "listAllTables" return "carbonTables" directly and the readLock is finished, "carbonTables" will be traversed without protection of lock 
so i return a clone of "carbonTables", it's thread-safe

**Comment by kumarvishal09** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2673/

**Comment by akashrn5** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10932/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2901/

**Comment by qiuchenjian** [comment]
retest this please

**Comment by qiuchenjian** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2675/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2903/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10934/

**Comment by qiuchenjian** [comment]
@kumarvishal09  @akashrn5   CI has passed without merge new code

**Comment by kumarvishal09** [comment]
LGTM

**Pull request closed without merging**